general/workspace.hpp

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) 2010-2024, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#ifndef MFEM_WORKSPACE_HPP
+#define MFEM_WORKSPACE_HPP
+
+#include "../linalg/vector.hpp"
+#include <forward_list>
+
+namespace mfem
+{
+
+// Forward declaration of internal::WorkspaceChunk
+namespace internal { class WorkspaceChunk; }
+
+/// @brief A vector used as a short-lived workspace for temporary calculations,
+/// created with Workspace::NewVector.
+///
+/// A WorkspaceVector is created using the Workspace bump allocator. The
+/// allocator can quickly and cheaply create new vectors, so that these vectors
+/// can be created and destroyed in loops without incurring the memory
+/// allocation and deallocation overhead.
+///
+/// WorkspaceVector%s should be used only for short-lived temporary storage; for
+/// example, they are not intended to be stored as member data in other classes.
+class WorkspaceVector : public Vector
+{
+   // using internal::WorkspaceChunk;
+   friend class internal::WorkspaceChunk;
+
+   /// The WorkspaceChunk containing the data for this vector.
+   internal::WorkspaceChunk &chunk;
+
+   /// Offset in the chunk.
+   const int offset;
+
+   /// Original size allocated.
+   const int original_size;
+
+   /// @brief Has this WorkspaceVector been moved from? If so, don't deallocate
+   /// from its WorkspaceChunk in the destructor.
+   bool moved_from = false;
+
+   /// Private constructor, create with Workspace::NewVector() instead.
+   WorkspaceVector(internal::WorkspaceChunk &chunk_, int offset_, int n);
+
+public:
+   /// @brief Move constructor. The moved-from WorkspaceVector has @a
+   /// size_in_chunk set to zero.
+   WorkspaceVector(WorkspaceVector &&other);
+
+   /// No copy constructor.
+   WorkspaceVector(WorkspaceVector &other) = delete;
```

Comment:
Why no `const` here and on the copy assignment?

Reply:
Here I don't think it matters (since it's deleted), but for the copy assignment it should be `const`.

Reply:
Fixed in [3c8c3ee](https://github.com/mfem/mfem/pull/4065/commits/3c8c3ee7f5e2b2b4ebcc0f6e6fd5c3952e10118b)

Reply:
> Here I don't think it matters (since it's deleted), but for the copy assignment it should be `const`.

But the `delete` would be removing the non-`const`, so i think as it was, there was still a `const` copy constructor.